packages/core/src/modules/oob/messages/OutOfBandMessage.ts

```diff
@@ -0,0 +1,123 @@
+import type { PlaintextMessage } from '../../../types'
+import type { DidCommService } from '../../connections/models/did/service/DidCommService'
+
+import { Expose, Type } from 'class-transformer'
+import { ArrayNotEmpty, Equals, IsArray, IsInstance, IsOptional, ValidateNested } from 'class-validator'
+import { parseUrl } from 'query-string'
+
+import { AgentMessage } from '../../../agent/AgentMessage'
+import { Attachment, AttachmentData } from '../../../decorators/attachment/Attachment'
+import { AriesFrameworkError } from '../../../error'
+import { JsonEncoder } from '../../../utils/JsonEncoder'
+import { JsonTransformer } from '../../../utils/JsonTransformer'
+import { MessageValidator } from '../../../utils/MessageValidator'
+
+interface OutOfBandMessageOptions {
+  id?: string
+  label?: string
+  goalCode?: string
+  goal?: string
+  accept?: string[]
+  handshakeProtocols?: string[]
+  services: DidCommService[]
+}
+
+export class OutOfBandMessage extends AgentMessage {
+  public constructor(options: OutOfBandMessageOptions) {
+    super()
+    if (options) {
+      this.id = options.id ?? this.generateId()
+      this.label = options.label
+      this.goalCode = options.goalCode
+      this.goal = options.goal
+      this.accept = options.accept
+      this.handshakeProtocols = options.handshakeProtocols
+      this.services = options.services
+    }
+  }
+
+  public addRequest(message: AgentMessage) {
+    if (!this.requests) this.requests = []
+    const requestAttachment = new Attachment({
+      id: this.generateId(),
+      mimeType: 'application/json',
+      data: new AttachmentData({
+        base64: JsonEncoder.toBase64(message.toJSON()),
+      }),
+    })
+    this.requests.push(requestAttachment)
+  }
+
+  public getRequests(): PlaintextMessage[] | undefined {
+    return this.requests?.map((request) => {
+      if (request.data.base64) {
+        return JsonEncoder.fromBase64(request.data.base64)
+      }
+
+      if (request.data.json) {
+        return request.data.json
+      }
+
+      throw new AriesFrameworkError('There is no supported data attribute in attachment')
+    })
```

Comment:
```suggestion
    return this.requests?.map((request) => request.data.getDataAsJson())
```

Reply:
That's great 👍  

FYI I was surprised a bit when I tried to call it as `request.getDataAsJson()`. Wouldn't be better to move the method up to `Attchment` class?

Reply:
Yeah sure, I think we can move it one level up!